src/Compilers/Core/Portable/MetadataReader/PEModule.cs

```diff
@@ -1623,26 +1621,109 @@ private bool HasStringAndIntValuedAttribute(EntityHandle token, AttributeDescrip
+            return false;
+        }
+
+        /// <summary>
+        /// Cracks the data on an Obsolete attribute using the Obsolete(string, bool) constructor.
+        /// </summary>
+        private static bool CrackObsoleteAttributeDataStringBool(out ObsoleteAttributeData? value, ref BlobReader sig)
         {
             string message;
             if (CrackStringInAttributeValue(out message, ref sig) && sig.RemainingBytes >= 1)
             {
                 bool isError = sig.ReadBoolean();
-                value = new ObsoleteAttributeData(ObsoleteAttributeKind.Obsolete, message, isError);
+
+                CrackObsoleteProperties(out var diagnosticId, out var urlFormat, ref sig);
+                value = new ObsoleteAttributeData(ObsoleteAttributeKind.Obsolete, message, isError, diagnosticId: diagnosticId, urlFormat: urlFormat);
                 return true;
             }
 
             value = null;
             return false;
         }
 
+        private static void CrackObsoleteProperties(out string? diagnosticId, out string? urlFormat, ref BlobReader sig)
+        {
+            diagnosticId = null;
+            urlFormat = null;
+
+            // See CIL spec section II.23.3 Custom attributes
+
+            // Next is a description of the optional “named” fields and properties.
+            // This starts with NumNamed– an unsigned int16 giving the number of “named” properties or fields that follow.
+            var numNamed = sig.ReadUInt16();
+            for (int i = 0; i < numNamed; i++)
+            {
+                // A NamedArg is simply a FixedArg preceded by information to identify which field or property it represents.
+                // PROPERTY is the single byte 0x54.
+                const int PROPERTY = 0x54;
+                var fieldOrProperty = sig.ReadByte();
+                if (fieldOrProperty != PROPERTY)
+                {
+                    return;
+                }
+
+                const int ELEMENT_TYPE_STRING = 0x0e;
+                var fieldOrPropType = sig.ReadByte();
+                if (fieldOrPropType != ELEMENT_TYPE_STRING)
+                {
+                    return;
+                }
+
+                // The FieldOrPropName is the name of the field or property, stored as a SerString.
+                if (!CrackStringInAttributeValue(out string fieldOrPropName, ref sig) ||
+                    !CrackStringInAttributeValue(out string fixedArg, ref sig))
+                {
+                    return;
+                }
+
+                switch (fieldOrPropName)
+                {
+                    case "DiagnosticId":
+                        diagnosticId = fixedArg;
+                        break;
+                    case "UrlFormat":
+                        urlFormat = fixedArg;
+                        break;
+                }
+
+                // TODO: do we want to tolerate custom definitions of ObsoleteAttribute with different properties types besides string?
```

Comment:
>TODO [](start = 19, length = 4)

TODO comments should be checked in

Reply:
Do you mean that I shouldn't commit such comments in my PRs, or that we shouldn't merge the PR until all such comments have been deleted?

Reply:
> Do you mean that I shouldn't commit such comments in my PRs, or that we shouldn't merge the PR until all such comments have been deleted?

We shouldn't merge PR with them.

---
In reply to: [395373780](https://github.com/dotnet/roslyn/pull/42518#discussion_r395373780) [](ancestors = 395373780)